ironcalc/IronCalc
# Merged Cells Implementation

### What are we trying to achieve with this PR ?

++ Currently IronCalc just imports and exports the merged cell without imposing rules of merged cells ( cell belonging to merge cells block cannot be modified etc ). This PR adds those conditions and restrictions which completes Merge cell implementation on IronCalc

### What changes are added in laymen terms ?

- New Merge cell struct has been added to cells.rs
- Check has been added to all public set fns where we check whether cell is being part of merged cells or not.
- Update and Delete merge cells ( Public API for handling merge cells ) have been added
- Exporting and importing of worksheet have been changed a bit to accommodate the above changes
- Testcase has been added
 

### What happens when user tries to create new merge cell which overlaps with other Merge cells ?

**Algorithm to Handle Overlapping Merges**

1.	Identify the Cells to Merge
Determine the range of cells you want to merge. For example, merge cells A1 and C1.
2.	Detect Existing Merged Areas
Check if any of the cells involved are part of existing merged cell ranges.
3.	Unmerge Overlapping Areas
Unmerge the areas where overlaps occur to avoid conflicts.
4.	Merge the New Range
Perform the merge operation on the new range.


### Manual testing and verification

case 1 : Merged cell in excell

![Screenshot 2024-10-09 135425](https://github.com/user-attachments/assets/5a529c0b-8c67-446e-aac1-f046e3a9dd99)

case 2 : Trying to create a overlapped merge cell ( C1 to G4 )

![Screenshot 2024-10-09 160344](https://github.com/user-attachments/assets/e7e1d6dd-97cb-493f-b11c-586a83fa4038)

++> In excell App Note that its not possible to select exactly C1 to G4, It will end up selecting all possible overlapped Merge cell

![Screenshot 2024-10-09 160430](https://github.com/user-attachments/assets/eae4678a-03cd-41db-a283-d03b65be3f6e)

++> Finally when you create new one is formed


**Same scenarios created in IronCalc**

case 1 : Merged cell

![Screenshot 2024-10-03 180501](https://github.com/user-attachments/assets/972f9954-229e-438b-8d52-04a19599fc3d)

case 2 : Trying to create overlapped merge cell ( C1 to G4 )

![Screenshot 2024-10-03 180544](https://github.com/user-attachments/assets/37ad86cc-1663-468b-8d63-a0c01c48c239)

++> Here its possible as we are creating using programming interface not user interface. But the logic remains same, where new merge cell to be created if overlaps with any of the existing merge cell, those gets unmerged and new one gets created
